master/internal/core_experiment.go

```diff
@@ -349,10 +349,24 @@ func (m *Master) parseCreateExperiment(ctx context.Context, req *apiv1.CreateExp
 		config.RawCheckpointStorage, &m.config.CheckpointStorage,
 	)
 
-	// Apply the scheduler's default priority.
+	// Apply the scheduler's default priority if priority is not set in an invariant config or		// constraint.
+	// constraint.
 	if config.Resources().Priority() == nil {
-		prio := masterConfig.DefaultPriorityForPool(poolName.String())
-		config.RawResources.RawPriority = &prio
+		// Returns an error if RM does not implement priority.
+		enforcedPriority, _, err := configpolicy.FindAllowedPriority(
```

Comment:
i'm confused why an unset priority becomes the `constraints.priority_limit`.

Reply:
Really good catch, this should be `enforcedPriority != nil`, rather than `enforcedPriority == nil`

Reply:
Oh wait, actually I think it's good as is!
This is basically saying that if no priority is enforced with config policies, then we use the default priority specified by the rm.
Otherwise, we use the enforced priority specified in the invariant config or constraints

Reply:
If you're talking about in the `FindAllowedPriority` function, I am not entirely sure where you mean 🙈

Reply:
what I'm saying is we want to constrain by the constraint not enforce it, right?

Reply:
I was thinking more that if a constraint is enforced and the user does not set priority in their experiment config (`config.Resources().Priority() == nil`),  then we just populate the priority with whatever the constraint is rather than rejecting the workload because the rm's default priority is higher than the constraint

Reply:
that might not be what we want though? 
@kkunapuli to weigh in on this!

Reply:
To the point about handling contradictions between defaults and constraints, I think we should reject the workload and admins should set sane defaults. I mentioned in my change this issue technically let's admins force end users to set a priority.

But my point about this code was it looks like if the priority is nil, we will set the priority to the enforcedPriority, which may be the constraints.priority_limit, even if the pool default and enforced priority are not in conflict.

Reply:
Very concretely, if I submit an experiment to `pool1`, don't pass a priority, there is a default of 42 and a constraint globally for 30, won't I end up with 30 not 42? Maybe I'm misreading.

Reply:
I see what you're saying. Yes you're correct that you end up with 30 and not 42. 
So you're saying we want to the 42 to be the "final" inputted priority and then the workload to get rejected due to a constraints violation so that users make sure to specifically set a constraint within the limit (otherwise we are just doing an inadvertent silent override with a constraint, which isn't desired)

if this is the case, I should probably close this PR

Reply:
now that im thinking about it, i do recall conversations where we talked about this and agreed we wouldn't do silent overrides with constraints

Reply:
> So you're saying we want to the 42 to be the "final" inputted priority and then the workload to get rejected due to a constraints violation so that users make sure to specifically set a priority within the limit

No in this case 42 > 30 so it would be above the limit, right?

Reply:
> No in this case 42 > 30 so it would be above the limit, right?

Depends on the RM. If it's an agentRM, it would be accepted and any priority value lower than the limit would be rejected.

If it's k8sRM, it would be rejected along with every priority value larger than the limit.

Reply:
Makes sense!
I just made the necessary changes

Reply:
cool. an aside: FindAllowedPriority feels strange; it's sort of leaking details about how it chose the priority. at least in this callsite I think a better API be `InvariantPriority(...) (prio int, ok bool)` that returns the invariant priority for a workload or (0, false) if there is none.

Reply:
curious what you think @kkunapuli

Reply:
> I think a better API be InvariantPriority(...) (prio int, ok bool) that returns the invariant priority for a workload or (0, false) if there is none.

+1 to this! I think we should separate out the logic for getting the invariant config and the constraints when finding different fields. `FindAllowedPriority` is _really_ well commented so it's possible to read the comments and understand the logic behind the return values, but I kinda feel like it's a bit too much logic packed into one func, particularly with the second output boolean.

For what it's worth, I have a [branch](https://github.com/determined-ai/determined/commit/7002d9cd4b3c17fca6986124ca6d40e8b70a243e) right now that uses generics and can potentially be used to shorten/refactor `FindAllowedPriority` to get the invariant configs in one foul swoop!

Reply:
@stoksc I agree that the function could use a refactor.

@amandavialva01 - the refactor should not impact functionality. With that in mind, please finish all other in-progress work before starting a refactor. Also, I would be more comfortable if we merge it *after* the release cutoff at this point.